docs/architecture/adr-008-app-caller-cbs.md

```diff
@@ -0,0 +1,111 @@
+
+E.g. Send an ICS-20 packet, and if it is successful, then send an ICA-packet to swap tokens on LP and return funds to sender
+
+This requires a second layer of callbacks. The IBC application already gets the result of the packet from core IBC, but currently there is no standardized way to pass this information on to a caller module/smart contract.
+
+## Decision
+
+Create a standardized callback interface that callers can implement. IBC applications (or middleware that wraps IBC applications) can now call this callback to route the result of the packet/channel handshake from core IBC to the IBC application to the original caller.
+
+IBC callers may implement the following interface:
+
+```go
+type IBCCaller interface {
+    // OnChannelOpen will be called on the IBCCaller when the channel opens
+    // this will happen either on ChanOpenAck or ChanOpenConfirm
+    OnChannelOpen(ctx sdk.Context, portID, channelID, version string)
+
+    // OnChannelClose will be called on the IBCCaller if the channel closes
+    // this will be called on either ChanCloseInit or ChanCloseConfirm and if the channel handshake fails on our end
+    // NOTE: currently the channel does not automatically close if the counterparty fails the handhshake so callers must be prepared for an OpenInit to never return a callback for the time being
+    OnChannelClose(ctx sdk.Context, portID, channelID string)
+
+    // OnAcknowledgementPacket will be called on the IBC Caller
+    // after the IBC Application handles its own OnAcknowledgementPacket callback
+    OnAcknowledgmentPacket(
+        ctx sdk.Context,
+        packet channeltypes.Packet,
+        ack exported.Acknowledgement
+    )
+
+    // OnTimeoutPacket will be called on the IBC Caller
+    // after the IBC Application handles its own OnTimeoutPacket callback
+    OnTimeoutPacket(
+        ctx sdk.Context,
+        packet channeltypes.Packet
+    )
+}
+```
+
+IBC Apps or middleware can then call the IBCCaller callbacks like so in their own callbacks:
+
+```go
+// this is the module-defined SendPacket function. It may differ from application to application
+// e.g. For ICS20 this would be the SendTransfer function
+func SendPacket(ctx sdk.Context, packet channeltypes.Packet, caller IBCaller) {
+    // do custom logic and send packet
+
+    // store a mapping of the packetID to the caller
+    k.setCaller(packetID, caller)
+}
+
+func OnAcknowledgementPacket(
+    ctx sdk.Context,
+    packet channeltypes.Packet,
+    acknowledgement []byte,
+    relayer string,
+) {
+    // application-specific onAcknowledgmentPacket logic
+
+    // unmarshal ack bytes into the acknowledgment interface
+    var ack exported.Acknowledgement
+    unmarshal(acknowledgement, ack)
+
+    // send acknowledgement to original caller
+    caller := k.getCaller(packetID)
+    caller.OnAcknowledgmentPacket(ctx, packet, ack)
```

Comment:
Would it be possible to do something like `deleteCaller(packetID)` if the packet lifecycle completes and we don't need the caller for this particular packet to be stored in state anymore? Maybe that would help with cleaning up? 

Same thing for the timeout callback...

Reply:
done in the new model